### PR TITLE
feat: Context Parallelism + Sequence Packing in Megatron + Dtensor

### PR DESCRIPTION
**Pull request by SahilJain314**
Enables Context Parallelism + Sequence Packing for Megatron-Core based training

image
Qwen 1.5B with TP + PP + CP (2,2,2) highlighted

### CODE REVIEW

File: nemo_rl/distributed/model_utils.py [hunk 1]
```diff
@@ -120,21 +120,21 @@ def backward(
         return grad_input, None, None, None, None, None, None
 
 
-def from_parallel_logits_to_logprobs(
-    vocab_parallel_logits: torch.Tensor,
-    target: torch.Tensor | DTensor,
+def dtensor_from_parallel_logits_to_logprobs(
```
Comment (terrykong) [inline nemo_rl/distributed/model_utils.py:123]: do you think you could add a test that tests `dtensor_from_parallel_logits_to_logprobs == from_parallel_logits_to_logprobs`? aside from being documentation, probably would expect these two to be equivalent provided certain equivalent inputs

File: nemo_rl/algorithms/loss_functions.py [hunk 2]
```diff
@@ -149,7 +150,9 @@ def __call__(
                 vocab_end_index=(vocab_parallel_rank + 1) * next_token_logits.shape[-1],
                 tp_group=vocab_parallel_group,
                 inference_only=False,
+                cp_group=context_parallel_group,
             )
+            curr_logprobs = curr_logprobs[:, : data["input_ids"].shape[1] - 1]
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py:156]: could you leave a comment on why this particular slicing is needed?

File: nemo_rl/distributed/model_utils.py [hunk 3]
```diff
@@ -188,3 +186,286 @@ def from_parallel_logits_to_logprobs(
...
+            probs, cp_group, seq_dim=1
+        )  # , unpadded_seqlen=target.shape[1])
+
+    if pad_len > 0:
+        probs = probs[:, :-pad_len]
+
+    return probs[:, :-1]
+
+
+def from_parallel_logits_to_logprobs_packed_sequences(
```
Comment (terrykong) [inline nemo_rl/distributed/model_utils.py:253]: similar Q about whether we can test `from_parallel_logits_to_logprobs_packed_sequences == from_parallel_logits_to_logprobs`

File: nemo_rl/models/megatron/common.py [hunk 4]
```diff
@@ -35,6 +255,11 @@ def forward_step_arbitrary_loss(
     data_iterator: Iterator[BatchedDataDict[Any]],
     model: GPTModel,
     loss_fn: LossFunction,
+    pack_sequences: bool = False,
+    seq_length_key: Optional[str] = None,
```
Comment (terrykong) [inline nemo_rl/models/megatron/common.py:259]: how often would you say this changes? any possibility it can just be assumed as a constant global var to reduce num args?

File: nemo_rl/models/megatron/common.py [hunk 5]
```diff
@@ -35,6 +255,11 @@ def forward_step_arbitrary_loss(
...
+    pad_individual_seqs_to_multiple_of: int = 1,
+    pad_full_seq_to: Optional[int] = None,
+    cp_normalize: bool = True,
```
Comment (terrykong) [inline nemo_rl/models/megatron/common.py:262]: I didn't see anywhere this is false, can this be inferred from the CP parallel group?

File: nemo_rl/models/policy/__init__.py [hunk 6]
```diff
@@ -117,7 +117,7 @@ class DynamicBatchingConfig(TypedDict):
     # training and logprob stages respectively.
     enabled: bool
     train_mb_tokens: int
-    logprob_mb_tokens: int
+    logprob_mb_tokens: NotRequired[int] = None
```
Comment (terrykong) [inline nemo_rl/models/policy/__init__.py]:
```suggestion
    logprob_mb_tokens: NotRequired[int]
```

File: nemo_rl/models/policy/lm_policy.py [hunk 7]
```diff
@@ -174,36 +201,54 @@ def get_logprobs(
...
-            sharded_data_2d.append(cp_data)
+        # sharded_data_2d = []
+        # shard_idx = 0
+        # # Convert to 2d dim array
+        # for _ in range(dp_size):
+        # cp_data = []
+        # for _ in range(cp_size):
+        # cp_data.append(sharded_data[shard_idx])
+        # shard_idx += 1
+        #     sharded_data_2d.append(cp_data)
```
Comment (terrykong) [inline nemo_rl/models/policy/lm_policy.py]: dbg code?

File: nemo_rl/distributed/batched_data_dict.py [hunk 8]
```diff
@@ -230,6 +248,19 @@ def shard_by_batch_size(
                                             4. input_lengths_key (str): the key in the batch
                                                 which holds the sequence length per value.
                                                 The sequence dim index is assumed to be 1.
+                                          Cannot be passed with sequence_packing_args.
+
+            sequence_packing_args (dict): If passed, preprocess batch for sequence packing. This
+                                            dict requires three keys:
```
Comment (terrykong) [inline nemo_rl/distributed/batched_data_dict.py]:
```suggestion
                                            dict requires five keys:
```

File: nemo_rl/distributed/batched_data_dict.py [hunk 9]
```diff
@@ -230,6 +248,19 @@ def shard_by_batch_size(
...
+                                            1. max_tokens_per_microbatch (int): the maximum
+                                                number of tokens in a microbatch
+                                            2. input_key (str): the key in the batch
+                                                which holds input ids.
+                                            3. input_lengths_key (str): the key in the batch
+                                                which holds the sequence length per value.
+                                                The sequence dim index is assumed to be 1.
+                                            4. algorithm (str): the algorithm to use for sequence packing.
```
Comment (terrykong) [inline nemo_rl/distributed/batched_data_dict.py:261]: is this understanding correct?
```suggestion
                                            4. algorithm (str): the algorithm to use for sequence packing.
                                            5. sequence_length_pad_multiple (int): sequence length to pad to. If CP or sequence parallel is enabled, this should be set to the least common multiple of CP and SP dimensions
```

File: nemo_rl/models/policy/lm_policy.py [hunk 10]
```diff
@@ -174,36 +201,54 @@ def get_logprobs(
                 batch_size=None,
                 dynamic_batching_args=self.dynamic_batching_args,
             )
+        elif self.use_sequence_packing:
+            self.sequence_packing_args["max_tokens_per_microbatch"] = self.cfg[
+                "sequence_packing"
+            ]["logprob_mb_tokens"]
+            sharded_data, unsorted_data_indices = data.shard_by_batch_size(
+                dp_size,
```
Comment (terrykong) [inline nemo_rl/models/policy/lm_policy.py:214]: could you add a comment on why sequence packing doesn't need to split by CP whereas the other branches do?

File: nemo_rl/models/policy/lm_policy.py [hunk 11]
```diff
@@ -230,28 +275,46 @@ def get_reference_policy_logprobs(
...
+        sharded_data_2d = sharded_data
+        # sharded_data_2d = []
+        # shard_idx = 0
+        # # Convert to 2d dim array
+        # for _ in range(dp_size):
+        # cp_data = []
+        # for _ in range(cp_size):
+        # cp_data.append(sharded_data[shard_idx])
+        # shard_idx += 1
+        # sharded_data_2d.append(cp_data)
```
Comment (terrykong) [inline nemo_rl/models/policy/lm_policy.py]: dbg comments?

File: nemo_rl/models/policy/lm_policy.py [hunk 12]
```diff
@@ -174,36 +201,54 @@ def get_logprobs(
...
             data=sharded_data_2d,
-            in_sharded_axes=["data_parallel", "context_parallel"],
-            replicate_on_axes=["tensor_parallel", "pipeline_parallel"],
-            output_is_replicated=["tensor_parallel", "pipeline_parallel"],
+            in_sharded_axes=["data_parallel"],
+            replicate_on_axes=[
+                "context_parallel",
+                "tensor_parallel",
+                "pipeline_parallel",
+            ],
```
Comment (terrykong) [inline nemo_rl/models/policy/lm_policy.py:232]: b/c the data is conditionally sharded different based on seqpacking on vs. (dynamic batch || neither), do these need to be conditional also on the same checks?

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 13]
```diff
@@ -796,11 +802,32 @@ def train(
...
+                    )
+                    data_iterator_len, seq_dim_size = (
+                        batch.get_microbatch_iterator_for_packable_sequences_len()
+                    )
+                    mbs = 1
+                    pack_seqs = True
+                    seqlen_key = "input_lengths"
+                    tp_size = self.cfg["megatron_cfg"]["tensor_model_parallel_size"]
+                    cp_size = self.cfg["megatron_cfg"]["context_parallel_size"]
+                    pad_factor = cp_size * 2 * tp_size if cp_size > 1 else tp_size
```
Comment (terrykong) [inline nemo_rl/models/policy/megatron_policy_worker.py:827]: for my education, where does this mysterious factor come from?

File: tests/unit/algorithms/test_sequence_packing_gradients.py [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+        # print("\n⚠️  This is a known issue being debugged. Test will pass but issue is documented.")
+        # else:
+        # print("✅ All gradients appear normal across all workers")
+
+        # # For now, we'll make the test pass even if high gradients are detected
+        # # so we can continue debugging. In a real scenario, you'd want to fix the issue.
+        # for i, result in enumerate(results):
+        # if not result["success"]:
+        # print(f"\n⚠️  Worker {i} detected high gradients - this is a known issue under investigation")
+        # # assert result["success"], f"Worker {i} failed: {result['error']}"
```
Comment (terrykong) [inline tests/unit/algorithms/test_sequence_packing_gradients.py]: is this supposed to be commented out?

File: tests/unit/algorithms/test_sequence_packing_gradients.py [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+
+        worker_group.shutdown(force=True)
+
+    finally:
+        cluster.shutdown()
+
+
+if __name__ == "__main__":
+    # Run the test directly
+    pytest.main([__file__, "-v", "-s"])
```
Comment (terrykong) [inline tests/unit/algorithms/test_sequence_packing_gradients.py]:
```suggestion
```

File: tests/unit/conftest.py [hunk 16]
```diff
@@ -335,8 +349,11 @@ def _distributed_test_wrapper(
...
         _cleanup_distributed()
     except Exception as e:
-        print(f"Error in rank {rank}: {e}")
-        _cleanup_distributed()
+        # report exception to the main process
+        import traceback
+
+        queue.put(f"Error in rank {rank}:\n{e}\n{traceback.format_exc()}")
+        # still raise the exception to stop the current process
         raise
```
Comment (terrykong) [inline tests/unit/conftest.py]: not sure i understand the necessity of the queue if it just errors out on the first and the above code only grabs the first thing from the queue to raise?

was the original implementation w/o the queue insufficient?

File: tests/unit/data/packing/test_algorithms.py [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+        # Check if the results are different
+        # This is a weak test, as randomness might still produce the same result
+        if sorted_bins1 == sorted_bins2:
+            print(
+                f"Warning: {algorithm.name} produced the same result with different seeds"
+            )
+
+
+if __name__ == "__main__":
+    pytest.main(["-xvs", __file__])
```
Comment (terrykong) [inline tests/unit/data/packing/test_algorithms.py]:
```suggestion
```

File: tests/unit/distributed/test_model_utils.py [hunk 18]
```diff
@@ -0,0 +1,426 @@
...
+    """Create a virtual cluster with 4 GPU bundles."""
+    cluster = RayVirtualCluster(bundle_ct_per_node_list=[4], use_gpus=True)
+    yield cluster
+    cluster.shutdown()
+
+
+import numpy as np
+
+
+@pytest.mark.unit
```
Comment (terrykong) [inline tests/unit/distributed/test_model_utils.py]:
```suggestion
```
i don't think we need this marker since we only use pytest for unit, but please let me know if you had some other reason to introduce

File: tests/unit/distributed/test_model_utils.py [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+    if ALLGATHER_CP_TEST_ACTOR_FQN in ACTOR_ENVIRONMENT_REGISTRY:
+        if original_registry_value is None:
+            del ACTOR_ENVIRONMENT_REGISTRY[ALLGATHER_CP_TEST_ACTOR_FQN]
+        else:
+            ACTOR_ENVIRONMENT_REGISTRY[ALLGATHER_CP_TEST_ACTOR_FQN] = (
+                original_registry_value
+            )
+
+
+@pytest.mark.unit
```
Comment (terrykong) [inline tests/unit/distributed/test_model_utils.py]:
```suggestion
```

File: tests/unit/models/megatron/test_common.py [hunk 20]
```diff
@@ -0,0 +1,709 @@
...
+            detailed = result["detailed_results"]
+            print(f"Worker {i} detailed results:")
+            for test_name, test_result in detailed.items():
+                status = "PASSED" if test_result["success"] else "FAILED"
+                print(f"  {test_name}: {status}")
+                if not test_result["success"]:
+                    print(f"    Error: {test_result['error']}")
+
+
+@pytest.mark.unit
```
Comment (terrykong) [inline tests/unit/models/megatron/test_common.py]:
```suggestion
```

File: tests/unit/models/megatron/test_common.py [hunk 21]
```diff
@@ -0,0 +1,709 @@
...
+
+    finally:
+        print("Cleaning up pack sequences test resources...")
+        if worker_group:
+            worker_group.shutdown(force=True)
+        if cluster:
+            cluster.shutdown()
+
+
+@pytest.mark.unit
```
Comment (terrykong) [inline tests/unit/models/megatron/test_common.py]:
```suggestion
```

File: tests/unit/algorithms/test_sequence_packing_gradients.py [hunk 22]
```diff
@@ -0,0 +1,488 @@
...
+    if SEQUENCE_PACKING_GRADIENT_TEST_ACTOR_FQN in ACTOR_ENVIRONMENT_REGISTRY:
+        if original_registry_value is None:
+            del ACTOR_ENVIRONMENT_REGISTRY[SEQUENCE_PACKING_GRADIENT_TEST_ACTOR_FQN]
+        else:
+            ACTOR_ENVIRONMENT_REGISTRY[SEQUENCE_PACKING_GRADIENT_TEST_ACTOR_FQN] = (
+                original_registry_value
+            )
+
+
+@pytest.mark.unit
```
Comment (terrykong) [inline tests/unit/algorithms/test_sequence_packing_gradients.py]:
```suggestion
```

**Review by terrykong — COMMENTED** [review]
thanks @ahmadki and @SahilJain314 

this is a big change so there are a lot of comments/questions :)

File: nemo_rl/models/megatron/common.py (hunk 4 above)
Comment (SahilJain314) [reply nemo_rl/models/megatron/common.py:259]: I'm generally against making anything a global var, but I think here in particular, these args are pretty important to this function so I'm happy with having it here.

File: nemo_rl/models/policy/megatron_policy_worker.py (hunk 13 above)
Comment (SahilJain314) [reply nemo_rl/models/policy/megatron_policy_worker.py:827]: It's to make sure we have a large enough number to divide into 2*CP*SP chunks. The 2 comes from the fact that CP chunking is actually chunked into 2*CP chunks so that we can stitch together chunks [0, N-1], [1, N-2] .... This does load balancing for causal masking (since it's triangular). The remaining factor is for SP.

File: nemo_rl/models/policy/lm_policy.py (hunk 12 above)
Comment (SahilJain314) [reply nemo_rl/models/policy/lm_policy.py:232]: These should be okay. The BatchedDataDict sharding call earlier only changes which sequences are in which microbatches, but doesn't actually perform any padding or packing (that's handled later)

File: nemo_rl/distributed/batched_data_dict.py (hunk 9 above)
Comment (SahilJain314) [reply nemo_rl/distributed/batched_data_dict.py:261]: should be 2*CP, but yes. Addressed in last commit.

File: nemo_rl/models/policy/megatron_policy_worker.py (hunk 13 above)
Comment (parthchadha) [reply nemo_rl/models/policy/megatron_policy_worker.py:827]: @SahilJain314 it would be nice to include that as a comment in the code for future readers.

**Comment by ashors1** [comment]
Quick comment -- can you make sure the nemo submodule contains all the [commits in main](https://github.com/NVIDIA/NeMo/tree/33259f2540af6eef375d43fc48bdcbd7ec490c29) (in particular, the qwen3 commit)? Just want to make sure we're not accidentally reverting anything

**Comment by SahilJain314** [comment]
I've reset nemo to the one on main. not sure why it was different in the first place...

File: nemo_rl/models/megatron/common.py (hunk 5 above)
Comment (SahilJain314) [reply nemo_rl/models/megatron/common.py:262]: I think this is good to leave as a flag for the user to configure if they want to. By default, I have it to true. This can't be inferred from the cp group as it controls whether we divide by CP to rescale the gradients correctly in the loss. It depends on if the implementation of CP gather would have grad flow I think, so it could change if the use does something different.

**Review by parthchadha — DISMISSED** [review]
